### PR TITLE
*: Add encryption config

### PR DESCRIPTION
**Pull request by yiwu-arbug**
###  What have you changed?
* Add config to enable encryption at rest.
* Improve encryption at rest error handling
* Cleanup previous usage of encrypted env

###  What is the type of the changes?
New Feature

###  How is the PR tested?
New unit tests and manual test.

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?
Yes

###  Does this PR affect `tidb-ansible`?
Need to update tikv config template.

### CODE REVIEW

File: components/encryption/src/encrypted_file/header.rs [hunk 1]
```diff
@@ -124,6 +99,23 @@ impl Header {
 
         buf.to_vec()
     }
+
+    pub fn verify(&self, content: &[u8]) -> Result<()> {
```
Comment (overvenus) [inline components/encryption/src/encrypted_file/header.rs]: We are no longer verify plaintext using header crc32.

File: components/encryption/src/encrypted_file/mod.rs [hunk 2]
```diff
@@ -31,25 +32,29 @@ impl<'a> EncryptedFile<'a> {
...
         let res = OpenOptions::new()
             .read(true)
             .open(self.base.join(self.name));
         match res {
             Ok(mut f) => {
                 let mut buf = Vec::new();
                 f.read_to_end(&mut buf)?;
-                let (_, content) = Header::parse(&buf)?;
+                let (header, content) = Header::parse(&buf)?;
+                println!("read encrypted len {}", content.len());
```
Comment (overvenus) [inline components/encryption/src/encrypted_file/mod.rs]: Please remove it.

File: components/encryption/src/encrypted_file/mod.rs [hunk 3]
```diff
@@ -31,25 +32,29 @@ impl<'a> EncryptedFile<'a> {
...
                 let mut encrypted_content = EncryptedContent::default();
                 encrypted_content.merge_from_bytes(content)?;
-                let plaintext = master_key.decrypt(&encrypted_content)?;
+                let plaintext = master_key
+                    .decrypt(&encrypted_content)
+                    .map_err(|e| Error::EncryptedFile(Box::new(e.compat())))?;
+                header.verify(&plaintext)?;
```
Comment (overvenus) [inline components/encryption/src/encrypted_file/mod.rs]: `Backend::decrypt` has already verified plaintext using sha256.

File: components/encryption/src/errors.rs [hunk 4]
```diff
@@ -18,6 +18,18 @@ pub enum Error {
...
+    #[fail(
+        display = "Current master key error {}, previous master key error {}.",
+        _0, _1
+    )]
+    DictDecrypt(
+        Box<dyn error::Error + Sync + Send>,
+        Box<dyn error::Error + Sync + Send>,
+    ),
+    #[fail(display = "Encrypted file error {}", _0)]
+    EncryptedFile(Box<dyn error::Error + Sync + Send>),
```
Comment (overvenus) [inline components/encryption/src/errors.rs]: Are they still necessary as we know whether the master is correct or not by checking sha256?

File: components/encryption/src/encrypted_file/mod.rs [hunk 5]
```diff
@@ -73,7 +78,8 @@ impl<'a> EncryptedFile<'a> {
             .encrypt(&plaintext_content)?
             .write_to_bytes()
             .unwrap();
-        let header = Header::new(&encrypted_content);
+        println!("write encrypted len {}", encrypted_content.len());
```
Comment (overvenus) [inline components/encryption/src/encrypted_file/mod.rs]: Please remove it.

File: components/encryption/src/master_key/file.rs [hunk 6]
```diff
@@ -4,27 +4,49 @@ use kvproto::encryptionpb::{EncryptedContent, EncryptionMethod};
...
-use crate::crypter::*;
-use crate::{AesCtrCrypter, Error, Iv, Result};
+use crate::crypter::{self, AesCtrCrypter, Iv};
+use crate::{Error, Result};
+
+use std::fs::File;
+use std::io::Read;
+use std::path::Path;
+
+use hex;
```
Comment (overvenus) [inline components/encryption/src/master_key/file.rs]: Can it be removed?

**Comment by yiwu-arbug** [comment]
still need more tests and adjuestment.

**Comment by overvenus** [comment]
Is this PR still in WIP?

**Comment by yiwu-arbug** [comment]
> Is this PR still in WIP?

Mostly done. Since there's quite some changes, I'll do another round of manual test before this is ready.

File: components/encryption/src/encrypted_file/mod.rs [hunk 7]
```diff
@@ -31,7 +31,7 @@ impl<'a> EncryptedFile<'a> {
         EncryptedFile { base, name }
     }
 
-    /// Read and decrypt the file.
+    /// Read and decrypt the file. Return none if file doesn't exist.
```
Comment (yiwu-arbug) [inline components/encryption/src/encrypted_file/mod.rs]: comment need update

**Comment by yiwu-arbug** [comment]
Remaining issues with this PR:
1. when change encryption method and restart, get rocksdb error about encryption key length incorrect. looks like wrong encryption method is passed to rocksdb in this case.
2. when turn off encryption (set method to plaintext) and setting `previous-master-key` but not `master-key`, tikv crash with protobuf error. It should correctly fallback in this case.
3. when dynamic config is enabled, tikv.toml is not being respect and default values of config is being used.

**Comment by yiwu-arbug** [comment]
Confirmed with @rleungx, dynamic-config-test is broken and will be fixed after https://github.com/pingcap/tidb-test/pull/1023

**Comment by yiwu-arbug** [comment]
Confirmed with @NingLin-P , the issue with dynamic config is due to https://github.com/pingcap/pd/pull/2282.

File: components/encryption/src/master_key/file.rs [hunk 8]
```diff
@@ -4,27 +4,60 @@ use kvproto::encryptionpb::{EncryptedContent, EncryptionMethod};
...
+                        )
+                        .into(),
+                    ));
+                }
+                let mut content = vec![];
+                let read_len = file.read_to_end(&mut content)?;
+                if read_len != file_len {
+                    return Err(Error::Other(
+                        format!(
+                            "mismatch master key file size read, expected {}, actual{}",
```
Comment (rleungx) [inline components/encryption/src/master_key/file.rs]:
```suggestion
                            "mismatch master key file size read, expected {}, actual {}",
```

File: components/encryption/src/master_key/file.rs [hunk 9]
```diff
@@ -4,27 +4,60 @@ use kvproto::encryptionpb::{EncryptedContent, EncryptionMethod};
...
+                            "mismatch master key file size, expected {}, actual {}.",
+                            key_len * 2 + 1,
+                            file_len,
+                        )
+                        .into(),
+                    ));
+                }
+                let mut content = vec![];
+                let read_len = file.read_to_end(&mut content)?;
+                if read_len != file_len {
```
Comment (hicqu) [inline components/encryption/src/master_key/file.rs]: Is it necessary? If length can't match, an error should be reported in `read_to_end`.

File: components/encryption/src/manager/mod.rs [hunk 10]
```diff
@@ -26,25 +27,67 @@ struct Dicts {
...
+                current_key_id: 0,
+                ..Default::default()
+            },
+            rotation_period,
+            base: Path::new(path).to_owned(),
+        }
+    }
+
+    fn open(
+        path: &str,
```
Comment (hicqu) [inline components/encryption/src/manager/mod.rs]: `path` is not necessary.

File: components/encryption/src/master_key/file.rs (hunk 9 above)
Comment (yiwu-arbug) [reply components/encryption/src/master_key/file.rs]: I briefly check `read_to_end` implementation and doesn't see it checks against file size. Doesn't hurt to be a bit paranoid.

**Comment by yiwu-arbug** [comment]
@overvenus @hicqu CI passes. Mind take another look?

**Pull request merged**